File: cssminifier.js
/*
	A simple CSS Minifier
*/

// Main function to minify CSS.

function cssminifier(css = ""){
	// Checking for the right type of argument.

	if (typeof css !== "string"){
		throw new Error("Invalid Type Passed.");
	}

	// Required Variables

	let minifiedcss = ``,
		inComment = false,
		spaceregex = /^[\s{}:]$/;

	for(let char = 0; char < css.length; char++){
		// Iterating over the css char by char.

		if(inComment === false){
			if(css[char] === '\n'){
				// Removing line breaks.
				continue;
			}
			else if(css[char] === '/' && css[char + 1] === '*'){
				// Start of a comment.
				inComment = true;	// We are inside a comment. Don't add anything unless it gets over.
				char++;				// No need to evaluate next character.
				continue;
			}
			else if(/\s/.test(css[char]) && (spaceregex.test(css[char + 1]) || spaceregex.test(css[char - 1]))){
				// Remove all the extra whitespaces.
				continue;
			}

			minifiedcss += css[char];
		}
		else{
			// If we are inside a comment. We don't need to add anything.
			// We just need to check if the comment has ended or not.

			if(css[char] === '*' && css[char + 1] === '/'){
				inComment = false;	// No longer inside a comment.
				char++;		// No need to evaluate next character.
				continue;
			}
			else{
				continue;	// Don't add the character as its inside a comment.
			}
		}
	}

	// Now removing all the instances where a closing bracket (}) comes after a semicolon as that isn't needed.

	minifiedcss = [...minifiedcss];	 // Turning the string into an array of characters.

	for(let char = 0; char < minifiedcss.length; char++){
		if(minifiedcss[char] === ';' && minifiedcss[char + 1] === '}'){
			minifiedcss[char] = '';	// No data in the character now.
		}
	}

	minifiedcss = minifiedcss.join('');	// Joining the array back to a string.

	return minifiedcss;
}

File: cssminifier.py
'''
	A Simple CSS Minifier.

	Author : Devesh Kumar.

	Process : 
	---------

	1. Get the CSS.
	2. Iterate over it character by character.
	3. Check if we are inside a comment. If yes, no need to add any characters inside the comment.
	4. If not, check if the character is a line break or an unnecessary whitespace. If yes, don't add it.
	5. If not, add the character to the minified css.
	6. Return the Minified CSS.
'''

# Importing Requirements

import re		# Regex Library

# Main Function to minify the CSS.

def cssminifier(css = ""):
	# Checking type of CSS
	if(type(css) is not str):
		print("\nInvalid Type CSS Passed.")
		exit()		# End execution

	# Required Variables

	minifiedcss = ''	# String that will be the Minified CSS.
	incomment = False	# Variable to keep track whether a comment is going on while iteration.
	spaceregex = "^[\s{}:;]$"	# Regex to check for unnecessary spaces.

	# Iterating over the CSS character by character.

	char = 0	# Iterator.

	while (char < len(css)):
		if(incomment == False):
			# If we are not inside a CSS comment.
			if(css[char] == '/'):
				if(char < len(css) - 1):
					if(css[char + 1] == '*'):
						# Start of a comment.
						incomment = True
						char += 1
						pass
			elif(css[char] == '\n'):
				# Line break.
				pass
			else:
				isspace = re.search("\s",css[char])
				nextisunnecessary = None
				previsunnecessary = None

				if(char < len(css) - 1 and char > 0):
					nextisunnecessary = re.search(spaceregex,css[char - 1])
					previsunnecessary = re.search(spaceregex, css[char + 1])

				if(isspace != None and (nextisunnecessary != None or previsunnecessary != None)):
					# The space is unnecessary
					pass
				else:
					minifiedcss += css[char]
		else:
			# If we are inside a comment.

			# Checking if the current character is a comment ender.

			if(css[char] == '*'):
				if(char < len(css) - 1):
					if(css[char + 1] == '/'):
						incomment = False	# No longer inside a comment.
						char += 1
						pass
			else:
				pass

		char += 1	# Keep incrementing character by character.

	# Now removing all the unnecessary semicolons from the string.

	minifiedcss = list(minifiedcss)

	char = 0

	while (char < len(minifiedcss)):
		if (char < len(minifiedcss) - 1):
			if(minifiedcss[char] == ';' and minifiedcss[char + 1] == '}'):
				minifiedcss[char] = ''	# Removing the block
		char += 1

	# Merging all the characters together.

	minifiedcss = ''.join(minifiedcss);

	# Returning the minifed CSS.

	return minifiedcss

''' 
# --------------------------------------------------------------
# Test Block : Remove the three single quotes above to run this.
# --------------------------------------------------------------

bigcss = """
/* This is a comment. */




body		{
	background : #ffffff;
	color :		blue;
    						}
.header{
	color : #fffff;
    }	
"""

print(cssminifier(bigcss))

'''